foxglove/extensions/toggle-joystick-panel/src/ToggleJoystickPanel.tsx

```diff
@@ -0,0 +1,358 @@
+  pitchAxis: number;
+  yawAxis: number;
+  torpedoActivate: boolean;
+  torpedoOneLaunch: boolean;
+  torpedoTwoLaunch: boolean;
+};
+
+type ToggleJoystickPanelState = {
+  error?: Error;
+  colorScheme?: RenderState["colorScheme"];
+  joystickEnabled: boolean;
+  transformedJoystickInputs: TransformedJoystickInputs;
+  joystickConnected: boolean;
+};
+
+function ToggleJoystickPanel({ context }: { context: PanelExtensionContext }): JSX.Element {
+  const [renderDone, setRenderDone] = useState<(() => void) | undefined>();
+  const [state, setState] = useState<ToggleJoystickPanelState>({
+    joystickEnabled: false,
+    joystickConnected: false,
+    transformedJoystickInputs: {
+      xAxis: 0,
+      yAxis: 0,
+      yawAxis: 0,
+      zAxis: 0,
+      pitchAxis: 0,
+      rollAxis: 0,
+      torpedoActivate: false,
+      torpedoOneLaunch: false,
+      torpedoTwoLaunch: false,
+    },
+  });
+
+  // Update color scheme
+  useEffect(() => {
+    context.onRender = (renderState: Immutable<RenderState>, done) => {
+      setState((oldState) => ({ ...oldState, colorScheme: renderState.colorScheme }));
+      setRenderDone(() => done);
+    };
+  }, [context]);
+  context.watch("colorScheme");
+
+  // Call our done function at the end of each render
+  useEffect(() => {
+    renderDone?.();
+  }, [renderDone]);
+
+  const toggleJoystick = useCallback(() => {
+    // Check if service calling is supported by the context
+    if (!context.callService) {
+      console.error("Calling services is not supported by this connection");
+      return;
+    }
+
+    // Request payload to toggle control types
+    const desiredControl: CustomMsgsControlTypesConst = state.joystickEnabled
+      ? CustomMsgsControlTypesConst.DESIRED_POSE
+      : CustomMsgsControlTypesConst.DESIRED_POWER;
+    const request: CustomMsgsSetControlTypesRequest = {
+      control_types: {
+        x: desiredControl,
+        y: desiredControl,
+        z: desiredControl,
+        roll: desiredControl,
+        pitch: desiredControl,
+        yaw: desiredControl,
+      },
+    };
+
+    // Make the service call
+    context.callService(SET_CONTROL_TYPES_SERVICE, request).then(
+      (response) => {
+        const typedResponse = response as CustomMsgsSetControlTypesResponse;
+
+        // Update the state based on the service response
+        // If the service responds with failure, display the response message as an error
+        if (typedResponse.success) {
+          setState((oldState) => ({ ...oldState, error: undefined, joystickEnabled: !oldState.joystickEnabled }));
+        } else {
+          setState((oldState) => ({ ...oldState, error: Error(typedResponse.message) }));
+        }
+      },
+      (error) => {
+        // Handle service call errors (e.g., service is not advertised)
+        setState((oldState) => ({ ...oldState, error: error as Error }));
+      },
+    );
+  }, [context, state.joystickEnabled]);
+
+  const publishPower = useCallback(
```

Comment:
Add a comment before this function in the same format as the one for `queryJoystick`.